### ISSUE TITLE
proposal: avoid_contextual_keywords_as_identifier

### ISSUE THREAD

**Issue by jacob314** [dart-lang/linter#3867]
## Description

Using contextual keywords as identifiers makes code confusing for users who think of many context keywords as reserved words.

## Details

For backwards compatability, Dart like other languages avoids adding new reserved words. Instead, they add "contextual keywords" or "built-in identifiers". These are identifiers that behave like keywords in some contexts but can be used by users as normal identifiers in other places.

For example, show behaves like a keyword when used after an import or export directive:
```dart
import 'foo.dart' show something;
```
But it can also be used as an identifier:

```dart
class Widget {
  void show() { print("I am visible now."); }
}
```

Generally using `show` as an identifierdoesn't confuse users but other contextual keywords like `async` and `when` is more confusing.
```dart
f(x) {
  switch (x) {
    case int when when when > 0:
      print('x is $when, which is greater than zero');
  }
}
```

## Kind

This lint prepares the way for future language evolution with [Escaped reserved words](https://github.com/dart-lang/language/issues/271)
Once escaped reserved words is implemented, this lint can be paired with a quickfix to escape contextual keywords that should be promoted to reserved words in the future.

## Bad Examples

```dart
f(x) {
  switch (x) {
    case int when when when > 0:
      print('x is $when, which is greater than zero');
  }
}

class when {}
f(x) {
  switch (x) {
    case when as as when when as as when == as as when:
      print('what?');
  }
}

class Foo {
  Future<int> async() async {
    return x;
  }
  int x;
}

class async {
 Future<async> async() async => this;
}

```

## Good Examples 

```dart
class Widget {
  void show() { print("I am visible now."); }
}

class Foo {
  Future<int> asyncValue() async {
    return x;
  }
  int x;
}

class Async {
 Future<Async> asyncValue() async => this;
}

```

## Discussion

Not all contextual keywords cause confusion. It might make sense to only prohibit contextual keywords (when, async) but not others (show). It could also make sense to only warn if the same name is used as both a context keyword and an identifier in the same file.
